libdevcrypto/LibSnark.cpp

```
@@ -0,0 +1,213 @@
+		BOOST_THROW_EXCEPTION(InvalidEncoding());
+	return p;
+}
+
+}
+
+pair<bool, bytes> dev::crypto::alt_bn128_pairing_product(dev::bytesConstRef _in)
+{
+	// Input: list of pairs of G1 and G2 points
+	// Output: 1 if pairing evaluates to 1, 0 otherwise (left-padded to 32 bytes)
+
+	size_t constexpr pairSize = 2 * 32 + 2 * 64;
+	size_t const pairs = _in.size() / pairSize;
+	if (pairs * pairSize != _in.size())
+	{
+		// Invalid length.
+		return {false, bytes{}};
+	}
+	try
+	{
+		initLibSnark();
+		libff::alt_bn128_Fq12 x = libff::alt_bn128_Fq12::one();
+		for (size_t i = 0; i < pairs; ++i)
+		{
+			dev::bytesConstRef pair = _in.cropped(i * pairSize, pairSize);
+			libff::alt_bn128_G2 p = decodePointG2(pair.cropped(2 * 32));
+			if (-libff::alt_bn128_G2::scalar_field::one() * p + p != libff::alt_bn128_G2::zero())
+				// p is not an element of the group (has wrong order)
+				return {false, bytes()};
+			if (p.is_zero())
+				continue; // the pairing is one
+			if (decodePointG1(pair).is_zero())
+				continue; // the pairing is one
+			x = x * libff::alt_bn128_miller_loop(
+				libff::alt_bn128_precompute_G1(decodePointG1(pair)),
+				libff::alt_bn128_precompute_G2(p)
+			);
+		}
+		bool result = libff::alt_bn128_final_exponentiation(x) == libff::alt_bn128_GT::one();
+		return {true, h256{result}.asBytes()};
+	}
+	catch (InvalidEncoding const&)
+	{
+		return {false, bytes{}};
+	}
+	catch (...)
+	{
+		cwarn << "Internal exception from libsnark. Forwarding as precompiled contract failure.";
+		return {false, bytes{}};
+	}
+}
+
+pair<bool, bytes> dev::crypto::alt_bn128_G1_add(dev::bytesConstRef _in)
+{
+	try
+	{
+		initLibSnark();
+		libff::alt_bn128_G1 const p1 = decodePointG1(_in);
+		libff::alt_bn128_G1 const p2 = decodePointG1(_in.cropped(32 * 2));
+		return {true, encodePointG1(p1 + p2)};
+	}
+	catch (InvalidEncoding const&)
+	{
+	}
+	catch (...)
+	{
+		cwarn << "Internal exception from libsnark. Forwarding as precompiled contract failure.";
```

Should we silence unexpected exceptions here?